src/shared/sleep-config.c

```diff
@@ -143,6 +146,136 @@ int read_battery_capacity_percentage(void) {
 
+/*Read system and battery identifier from specific location and generate hash of it*/
+static int get_system_battery_identifier_hash(uint8_t ret[static SHA256_DIGEST_SIZE]) {
+        _cleanup_free_ char *bat_manufacturer = NULL, *bat_model_name = NULL, *bat_serial_number = NULL;
+        sd_id128_t machine_id, product_id;
+        struct sha256_ctx hash;
+        int r;
+
+        sha256_init_ctx(&hash);
+        r = read_one_line_file("/sys/class/power_supply/BAT0/manufacturer", &bat_manufacturer);
+        if (r > 0)
+               sha256_process_bytes(bat_manufacturer, strlen(bat_manufacturer), &hash);
+        else if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/manufacturer is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/manufacturer: %m");
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/model_name", &bat_model_name);
+        if (r > 0)
+               sha256_process_bytes(bat_model_name, strlen(bat_model_name), &hash);
+        else if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/model_name is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/model_name: %m");
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/serial_number", &bat_serial_number);
+        if (r > 0)
+               sha256_process_bytes(bat_serial_number, strlen(bat_serial_number), &hash);
+        else if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/serial_number is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/serial_number: %m");
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r > 0)
+               sha256_process_bytes(&machine_id, sizeof(machine_id), &hash);
+        else if (r == -ENOENT)
+               log_debug_errno(r, "machine ID is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to get machine ID: %m");
+
+        r = id128_get_product(&product_id);
+        if (r > 0)
+               sha256_process_bytes(&product_id, sizeof(product_id), &hash);
+        else if (r == -ENOENT)
+               log_debug_errno(r, "product_id does not exist: %m");
+        else if (r < 0)
+                return log_debug_errno(r, "Failed to get product ID: %m");
+
+        sha256_finish_ctx(&hash, ret);
+
+        return 0;
+}
+
+/* battery percentage discharge rate per hour is in range 0-200 then return success */
+static int battery_discharge_rate_in_range(int battery_discharge_rate) {
+        if (battery_discharge_rate <= 0 || battery_discharge_rate >= 200)
+               return 0;
+               /* battery discharge rate is not in a valid range */
+        log_debug("Battery discharge rate is in valid range");
+        return 1;
```

Comment:
This is a tiny helper function, hence I think debugging log should not necessary:
```
bool battery_discharge_rate_is_valid(int rate) {
        return rate >= 0 && rate <= 200;
}
```
Also, the comment says that rate can be 0 and 200...

Reply:
Yes. will remove that on final push. Just added to have verbose log for my convenience while debugging.

Reply:
Also will it not be implicit cast if i change the type of function ?
if (!battery_discharge_rate_is_valid(discharge_rate))

Reply:
Right. By changing the return type to `bool`, then you can write so.

Reply:
great. fixed. Just removed the equal sign. can't store discharge rate 0 because that is invaluable data and 200 means battery will discharge in 30 mins that is again the extreme since we are going to keep a buffer of 30 min.